example/bench/parameters.cpp

```diff
@@ -0,0 +1,111 @@
+#include <exception>
+#include <fstream>
+#include <string>
+
+#include <util/optional.hpp>
+
+#include <json/json.hpp>
+
+#include "parameters.hpp"
+
+double bench_params::expected_advance_time() const {
+    return cell.realtime_ratio * duration*1e-3 * num_cells;
+}
+unsigned bench_params::expected_spikes() const {
+    return num_cells * duration*1e-3 * cell.spike_freq_hz;
+}
+unsigned bench_params::expected_spikes_per_interval() const {
+    return num_cells * network.min_delay*1e-3/2 * cell.spike_freq_hz;
+}
+unsigned bench_params::expected_events() const {
+    return expected_spikes() * network.fan_in;
+}
+unsigned bench_params::expected_events_per_interval() const {
+    return expected_spikes_per_interval() * network.fan_in;
+}
+
+std::ostream& operator<<(std::ostream& o, const bench_params& p) {
+    o << "benchmark parameters:\n"
+      << "  name:          " << p.name << "\n"
+      << "  num cells:     " << p.num_cells << "\n"
+      << "  duration:      " << p.duration << " ms\n"
+      << "  fan in:        " << p.network.fan_in << " connections/cell\n"
+      << "  min delay:     " << p.network.min_delay << " ms\n"
+      << "  spike freq:    " << p.cell.spike_freq_hz << " Hz\n"
+      << "  cell overhead: " << p.cell.realtime_ratio << " ms to advance 1 ms\n";
+    o << "expected:\n"
+      << "  cell advance: " << p.expected_advance_time() << " s\n"
+      << "  spikes:       " << p.expected_spikes() << "\n"
+      << "  events:       " << p.expected_events() << "\n"
+      << "  spikes:       " << p.expected_spikes_per_interval() << " per interval\n"
+      << "  events:       " << p.expected_events_per_interval()/p.num_cells << " per cell per interval";
+    return o;
+}
+
+using arb::util::optional;
+using nlohmann::json;
+
+// Search a json object for an entry with a given name.
+// If found, return the value and remove from json object.
+template<typename T>
+optional<T> extract(const char* name, json& j) {
+    auto it = j.find(name);
+    if (it==j.end()) {
+        return arb::util::nullopt;
+    }
+    T value = std::move(*it);
+    j.erase(name);
+    return std::move(value);
+}
+
+bench_params read_options(int argc, char** argv) {
+    bench_params params;
+    if (argc<2) {
+        std::cout << "Using default parameters.\n";
+        return params;
+    }
+    if (argc>2) {
+        throw std::runtime_error("More than command line one option not permitted.");
+    }
+
+    std::string fname = argv[1];
+    std::cout << "Loading parameters from file: " << fname << "\n";
+    std::ifstream f(fname);
+
+    if (!f.good()) {
+        throw std::runtime_error("Unable to open input parameter file: "+fname);
+    }
+
+    nlohmann::json json;
+    json << f;
+
+    if (auto o  = extract<std::string>("name", json)) {
```

Comment:
So, none of these parameters can be something equivalent to false? No empty strings, no 0 num-cells, no 0 duration, ...? Do we have that documented somewhere?

Reply:
This code will do a reasonable job of testing that the correct type was used for each parameter, but you are right, it doesn't then check that the parameters are in valid ranges.

This is intended to be simple example code, so I was trying to tread the line between readability and a reasonable level of robustness (hence the efforts to warn users about unused parameters, which would normally indicate a typo).